### ISSUE TITLE
Stream.rotate does not work for inconsistent data gaps

### ISSUE THREAD

**Issue by iandkelly** [obspy/obspy#3462]
### Avoid duplicates

- [X] I searched existing issues

### Bug Summary

I'm trying to rotate some daily ambient noise data downloaded (via MassDownloader) for a particular station (IU.QSPA.00.BH*) from station coordinates to ZNE.

I made a simple function (as below) to retrieve all channel waveforms from a local directory, check and join them into a 3-C Stream object if all 3 channels are downloaded for a particular day, and then rotate this Stream and rewrite back out each channel trace as an .mseed file into a new directory.

However, an error message popped up (as below) - looking into what caused the issue, it seems like theres an arbitrary condition to force the traces in Stream objects during rotation to be a multiple of 3, which caused the issue when a certain day had inconsistent data gaps between channels, eg.,

> 7 Trace(s) in Stream:
IU.QSPA.00.BH1 | 2009-01-21T00:00:00.041000Z - 2009-01-21T01:03:42.891000Z | 20.0 Hz, 76458 samples
IU.QSPA.00.BH1 | 2009-01-21T01:57:23.740237Z - 2009-01-21T23:59:59.940237Z | 20.0 Hz, 1587125 samples
IU.QSPA.00.BH1 | 2009-01-21T01:04:13.841567Z - 2009-01-21T01:57:23.691567Z | 20.0 Hz, 63798 samples
IU.QSPA.00.BH2 | 2009-01-21T00:00:00.041000Z - 2009-01-21T01:03:32.341000Z | 20.0 Hz, 76247 samples
IU.QSPA.00.BH2 | 2009-01-21T01:04:02.940567Z - 2009-01-21T23:59:59.940567Z | 20.0 Hz, 1651141 samples
IU.QSPA.00.BHZ | 2009-01-21T00:00:00.040900Z - 2009-01-21T01:03:27.440900Z | 20.0 Hz, 76149 samples
IU.QSPA.00.BHZ | 2009-01-21T01:03:55.490567Z - 2009-01-21T23:59:59.940567Z | 20.0 Hz, 1651290 samples

I've just merged the data gaps for problematic days to avoid this issue, but it seems like this condition shouldn't be needed if the Stream still contains 3-C traces pertaining to the same station? 

Apologies if I've missed a duplicate of this!

### Code to Reproduce

```python
def RotateWaveforms(network, station, location, channels, waveform_dir, rotated_dir):  
  """
  Function to rotate all waveforms for a particular station found within a local directory via Obspy's Stream.rotate.

  Parameters
  ----------
  network : str
      SEED network code for station.

  station : str
      SEED station code.

  location : str
      SEED location code for station.

  channels : str
      Code for station channels - please put as wildcard (eg., 'BH*').

  waveform_dir : str
      Full filepath to directory where raw waveform data is stored for station.

  rotated_dir : dir
      Full filepath to directory where rotated waveform data will be stored.

  Returns
  -------
  None (waveforms are automatically written out).
  """

  # Raise error if channels is not set as wildcard
  if not channels.endswith('H*'):
    raise ValueError('Channels not set correctly! Please ensure channels is given as a wildcard, e.g., BH*.')

  # Also raise error if waveform_dir is not or incorrectly provided
  if (waveform_dir is None) or (not os.path.exists(waveform_dir)):
    raise ValueError('Missing or invalid waveform directory!')

  # Also check rotated_dir
  if not os.path.exists(rotated_dir):
    os.makedirs(rotated_dir)

  # Define station ID for convenience
  station_id = '{n}.{s}.{l}.{c}'.format(n = network, s = station, l = location, c = channels)

  # Retrieve all waveform files in directory
  files = glob.glob(os.path.join(waveform_dir, station_id + '__*.mseed'))
  if len(files) == 0:
    raise FileNotFoundError('No waveforms found! Please check if waveform_dir is provided correctly, or if specified station corresponds to specified waveform_dir.')

  # Then need to make sure only to select/rotate daily waveforms available for all 3 components 
  # So first compile waveforms for each day available/downloaded
  waveform_dict = {}
  for file in files:
    starttime = file.split('__')[1]
    endtime = file.split('__')[2].split('.')[0]
    waveform_day = starttime + '__' + endtime
    if waveform_day not in waveform_dict:
      waveform_dict[waveform_day] = [file]
    else:
      waveform_dict[waveform_day].append(file)
  # Then remove days where all 3 components aren't downloaded
  days_to_remove = [day for day in waveform_dict if len(waveform_dict[day]) != 3]
  for day in days_to_remove:
    waveform_dict.pop(day, None)

  # Get Inventory object for station
  inv_channels = ','.join(channels.replace('*', comp) for comp in ['1', '2', 'Z']) # N.B. get_stations needs channels provided as comma-separated string
  inv = Client('IRIS').get_stations(network = network, station = station, location = location, channel = inv_channels, level = 'channel')

  warnings.filterwarnings('ignore', category = UserWarning)

  # Rotate each 3-C waveform and rewrite out separate as component traces
  for day, trace_list in waveform_dict.items():
    stream = Stream()
    for trace in trace_list:
      stream += read(trace)
    rotated_stream = stream.rotate(method = '->ZNE', inventory = inv)
      for trace in rotated_stream:
        rotated_filename = '{n}.{s}.{l}.{c}__{t}.mseed'.format(n = trace.stats.network, s = trace.stats.station, l = trace.stats.location, c = trace.stats.channel, t = day)
        rotated_filepath = os.path.join(rotated_dir, rotated_filename)
        trace.write(rotated_filepath, format = 'MSEED')

  return
```

### Error Traceback

```Python traceback
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[1], line 3
----> 3 RotateWaveforms(network = 'IU',
      4                       station = 'QSPA',
      5                       location = '00',
      6                       channels = 'BH*',
      7                       waveform_dir = '/Volumes/LaCie/SeismicData/IU/QSPA/waveforms',
      8                       rotated_dir = '/Volumes/LaCie/SeismicData/IU/QSPA/waveforms_rotated')

File ~/Documents/UTAS/PhD/PhD_HVSR/Coding/PhD_HVSR_Utils.py:980, in RotateWaveforms(network, station, location, channels, waveform_dir, rotated_dir)
    978 for trace in trace_list:
    979   stream += read(trace)
--> 980 rotated_stream = stream.rotate(method = '->ZNE', inventory = inv)
    981 for trace in rotated_stream:
    982   rotated_filename = '{n}.{s}.{l}.{c}__{t}.mseed'.format(n = trace.stats.network, s = trace.stats.station, l = trace.stats.location, c = trace.stats.channel, t = day)

File ~/micromamba/envs/myhvsrpy/lib/python3.11/site-packages/obspy/core/stream.py:2786, in Stream.rotate(self, method, back_azimuth, inclination, inventory, **kwargs)
   2783         msg = ("With method '->ZNE' station metadata has to be "
   2784                "provided as 'inventory' parameter.")
   2785         raise ValueError(msg)
-> 2786     return self._rotate_to_zne(inventory, **kwargs)
   2787 elif method == "NE->RT":
   2788     func = "rotate_ne_rt"

File ~/micromamba/envs/myhvsrpy/lib/python3.11/site-packages/obspy/core/stream.py:3532, in Stream._rotate_to_zne(self, inventory, components)
   3529             if components == set(component_pair):
   3530                 channels_ = [cha_without_comp + comp
   3531                              for comp in component_pair]
-> 3532                 self._rotate_specific_channels_to_zne(
   3533                     net, sta, loc, channels_, inventory)
   3534 return self

File ~/micromamba/envs/myhvsrpy/lib/python3.11/site-packages/obspy/core/stream.py:3595, in Stream._rotate_specific_channels_to_zne(self, network, station, location, channels, inventory)
   3592 if len(st) % 3 != 0:
   3593     msg = ("Unexpected behavior in rotation. Please file a bug "
   3594            "report on github.")
-> 3595     raise NotImplementedError(msg)
   3596 num_pieces = int(len(st) / 3)
   3597 for i in range(num_pieces):
   3598     # three consecutive traces are always the ones that combine for one
   3599     # rotation run

NotImplementedError: Unexpected behavior in rotation. Please file a bug report on github.
```

### ObsPy Version?

1.4.0

### Operating System?

OS Sonoma 14.5

### Python Version?

3.11

### Installation Method?

conda

**Comment by megies** [comment]
The reason you run into trouble is that one of the channels has a gap that is 2.6% of the sample spacing. There is a [cleanup merge](https://docs.obspy.org/master/packages/autogen/obspy.core.stream.Stream.html#obspy.core.stream.Stream._cleanup) done internally but the default misalignment to merge traces together is only 1% of the sample spacing. That default might be a bit low, but this is quite the rare case what you got there. I have not ever seen a digitizer before that would decide to adjust to clock timing when it's timing is off by 2% of the sampling interval.

Basically what happens if you make that threshold larger, the second part of the trace will get shifted on subsample scale (only the times of the samples get shifted, no real interpolation or anything) so that it perfectly aligns with the first part.

What you can do is simply manually add a cleanup merge that allows for a larger misalignment:

```python
st.merge(method=-1, misalignment_threshold=0.1)
st.rotate(...)
```

**Comment by megies** [comment]
I think we should just add a warning to be shown or even a `ValueError` if `Stream._trim_common_channels()` ends up being unable to do its job due to too large misalignment.
Or maybe when that happens show a warning that the misalignment was larger than the default allowed threshold and then crank up the threshold and redo the trimming.

**Comment by megies** [comment]
I also find it quite odd, that the three channels have their timing shifted on sub sample scale. Why would the digitizer decide to use different sample timing for the different channels it's sampling.

**Comment by iandkelly** [comment]
Ah okay, I didn't spot that. 

So just to confirm then - would this be the only reason why this error is produced? ie., you wouldn't expect inconsistent data gaps between channels otherwise?

**Comment by megies** [comment]
> would this be the only reason why this error is produced? ie., you wouldn't expect inconsistent data gaps between channels otherwise?

Well, I didn't expect this one, so not sure what to answer there, if there was something I'd expect to make problems, I'd already work around it :smile: 
I believe in general it works fine, or we'd get more reports with problems since I believe rotating to ZNE is probably one of the more heavily used operations out there.

**Comment by ThomasLecocq** [comment]
>I also find it quite odd, that the three channels have their timing shifted on sub sample scale. Why would the digitizer decide to use different sample timing for the different channels it's sampling.

Ask Reftek :-) I only encoutered that kind of timing with their data...

**Comment by megies** [comment]
If I remember correctly, there is some workarounds users can do if they have the problem and find this issue, and there's no PR being worked on yet, so we'll have to bump this to next release
